[WIP] Cache FilesystemView early and reuse to reduce listStatus calls on namenode

bvaradar opened the pull request:
Issue : https://github.com/uber/hudi/issues/433

    Hudi writing and Compaction step to use cached file-system view containing only needed file-slices to reduce name-node list-status calls
    Write Path :
         When using Bloom Filter index, listStatus is performed once and cached in BloomIndex for all interesting partition-paths. For other cases, listStatus is performed after records are prepped (indexed)
         and before starting the IOHandle write stage. This helps avoid listStatus calls for each IOHandle (MergeHandle and AppendHandle) tasks. Depending on the partition-to-file distribution, the gains vary.
         In case where all the fileIds to be merged/appended are in the same partition-path, we save n-1 listStatus calls where n is the number of fileIds changed. In case where only new parquet-files are generated
         or the corner-case where there is exactly one fileId per partition getting appended/merged, #listStatus calls are break-even for bloom index.
    Compaction Path:
         Hoodie Compactor caches all the partitions which has atleast one fileId in the scheduled compaction plan. The benefits are similar to above except that only MergeHandle is used here which always does one listStatus
         at partition-level. For cases where there are more than one fileId per partition getting compacted, we only do listStatus once per partition and save others.

bvaradar commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieDataFile.java:
```diff
@@ -18,53 +18,53 @@
-import java.util.Comparator;
 import org.apache.hadoop.fs.FileStatus;
 
 public class HoodieDataFile implements Serializable {
```
Made HoodieFileGroup and its containing objects (HoodieDataFile/HoodieLogFile) serializable

bvaradar commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieLogFile.java:
```diff
@@ -38,49 +39,53 @@
   public HoodieLogFile(FileStatus fileStatus) {
     this(fileStatus.getPath());
-    this.fileStatus = Optional.of(fileStatus);
+    this.fileLen = com.google.common.base.Optional.of(fileStatus.getLen());
```
Guava Optional is serializable. To avoid cascadiing changes, have limited the change from java.util.Optional to Guava Optional to only needed places

bvaradar commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/TableFileSystemView.java:
```diff
@@ -30,9 +30,9 @@
+   * ReadOptimizedView with methods to only access latest version of file for the instant(s) passed.
    */
-  interface ReadOptimizedView {
+  interface LatestVersionOnlyReadOptimizedView {
```
Limiting the API usable by LatestFileSliceOnlyFSView

vinothchandar replied:
rename to `ReadOptimizedViewWithLatestSlice`?

bvaradar commented:
@n3nash @vinothchandar : when you get a chance, please review this.

n3nash commented:
@bvaradar Before reviewing the diff, have one question. If the intention of the diff to reduce listStatus calls when accessing the filesystem view repeatedly within the same job or is the intention to reduce across ingestion vs compaction ?

bvaradar commented:
@n3nash : The intent was to avoid the repeated list-status calls from each IO handle task during ingestion and compaction. No change on the reader's path though. This is similar to the closed PR https://github.com/uber/hudi/pull/278

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/table/HoodieTable.java:
```diff
@@ -120,6 +124,17 @@ public TableFileSystemView getFileSystemView() {
+   * Get the read optimized view of the file system for this table
+   */
+  public SealedLatestVersionOnlyFSView getLatestFileSliceOnlyFSView() {
+    if (null == latestFileSliceOnlyFSView) {
```
just a thought.. do we need to protect this synchronized? probably not, since driver access would be single threaded anyway.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieDataFile.java:
```diff
@@ -18,53 +18,53 @@
 
-  private FileStatus fileStatus;
+  // FileStatus not serializable in HDP 2.x (See: https://issues.apache.org/jira/browse/HADOOP-13895)
+  private transient FileStatus fileStatus;
```
even when using kryo?

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieFileGroup.java:
```diff
@@ -33,10 +33,7 @@
-      // reverse the order
-      return o2.compareTo(o1);
-    };
+    return Comparator.reverseOrder();
```
good cleanup

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieFileGroup.java:
```diff
@@ -62,14 +59,16 @@
     this.timeline = timeline;
-    this.lastInstant = timeline.lastInstant();
+    // Guava Optional is serializable.
+    // TODO : varadarb -> Open another PR to replace java.util.Optional with Guava'a everywhere internally
```
Open to trying it.. but Guava is tricky.. it usually has ton of version conflicts.. Roll our own instead?

n3nash replied:
I have only one concern with using our own Optional type -> If any of these classes are exposed to external users, then for them to understand a custom implementation of Optional is an overhead. Most people are used to standard Guava or Java optionals...

bvaradar replied:
Went ahead and made this change.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieFileGroup.java:
```diff
@@ -62,14 +59,16 @@
    * The last completed instant, that acts as a high watermark for all getters
    */
-  private final Optional<HoodieInstant> lastInstant;
+  private final com.google.common.base.Optional<HoodieInstant> lastInstant;
```
who else uses Optional in this file? could nt we import guava optional and avoid explict package naming in definition?

n3nash replied:
+1

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/TableFileSystemView.java:
```diff
@@ -45,34 +45,38 @@
+   * RealtimeView with methods to only access latest version of file-slice for the instant(s) passed.
    */
-  interface RealtimeView {
+  interface LatestVersionOnlyRealtimeView {
```
similar rename here

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/view/LatestFileSliceOnlyFSViewImpl.java:
```diff
@@ -0,0 +1,78 @@
+
+  protected HoodieFileGroup preProcessFileGroup(HoodieFileGroup group, boolean isCompactionPending) {
+    HoodieFileGroup newGroup = new HoodieFileGroup(group.getPartitionPath(), group.getId(), group.getTimeline());
+    int limit = 1;
```
`int limit = isCompactionPending ? 2 : 1` ?

n3nash replied:
s/limit/maxNumberOfFileSlicesAdded ?

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:
```diff
@@ -423,6 +423,14 @@ private void saveWorkloadProfileMetadataToInflight(WorkloadProfile profile, Hood
     }
 
+    // Cache Latest File-System View & Seal
+    preppedRecords.map(rec -> rec.getPartitionPath()).distinct().collect().stream().forEach(partitionPath -> {
```
we should probably test and ensure this addtional `distinct` is okay.. moving this listing out of index does seem like the right/cleaner this to do btw.. (a serious concern I had with the old PR)

n3nash replied:
This is introducing another action on the driver and collecting all records right ? Even after the records are persisted, this may introduce additional overhead.

bvaradar replied:
@vinothchandar @n3nash : Added the config with default false. We can discuss f2f on how to measure #liststatus calls.

n3nash replied:
sounds good

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:
```diff
@@ -423,6 +423,14 @@ private void saveWorkloadProfileMetadataToInflight(WorkloadProfile profile, Hood
+    // Cache Latest File-System View & Seal
+    preppedRecords.map(rec -> rec.getPartitionPath()).distinct().collect().stream().forEach(partitionPath -> {
+      // Populate latest version file-slice cache. No-op if already populated
+      long numFileSlices = hoodieTable.getLatestFileSliceOnlyFSView().getLatestFileSlices(partitionPath).count();
```
How does the populated values on executor make its way back to driver?

n3nash replied:
+1

n3nash replied:
ohh I see, the collect() is before you perform the getLatestFileSlice map operations.

bvaradar replied:
I am moving this part to a separate utility class to avoid confusion and be reusable across places.

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleanHelper.java:
```diff
@@ -102,7 +102,7 @@ public HoodieCleanHelper(HoodieTable<T> hoodieTable, HoodieWriteConfig config) {
           if (nextSlice.getDataFile().isPresent()) {
             HoodieDataFile dataFile = nextSlice.getDataFile().get();
-            deletePaths.add(dataFile.getFileStatus().getPath().toString());
+            deletePaths.add(dataFile.getPath());
```
unrelated to code here..  but cleaner will list the partitions again correct?

vinothchandar replied:
i.e does nt share the cache?

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java:
```diff
@@ -60,7 +60,7 @@
   private Set<String> writtenRecordKeys;
   private HoodieStorageWriter<IndexedRecord> storageWriter;
-  private TableFileSystemView.ReadOptimizedView fileSystemView;
+  private TableFileSystemView.LatestVersionOnlyReadOptimizedView fileSystemView;
```
So create handle does not need any changes?

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieDataFile.java:
```diff
@@ -18,53 +18,53 @@
+  // FileStatus not serializable in HDP 2.x (See: https://issues.apache.org/jira/browse/HADOOP-13895)
+  private transient FileStatus fileStatus;
+  private final String fileName;
+  private final String fullPathStr;
```
Data type is String already, so "fullPath" is enough ?

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieDataFile.java:
```diff
@@ -18,53 +18,53 @@
 
   public HoodieDataFile(FileStatus fileStatus) {
     this.fileStatus = fileStatus;
+    this.fileName = fileStatus == null ? null : fileStatus.getPath().getName();
```
To check if a file status exists for a HoodieDataFile, we will have to check against each of these fields == null, should we add a method hasFileStatus() or something where you can perform this null check internally and then provide boolean to the user of this class ?

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieFileGroup.java:
```diff
@@ -62,14 +59,16 @@
-    this.lastInstant = timeline.lastInstant();
+    // Guava Optional is serializable.
+    // TODO : varadarb -> Open another PR to replace java.util.Optional with Guava'a everywhere internally
+    this.lastInstant = com.google.common.base.Optional.fromNullable(timeline.lastInstant().orElse(null));
```
Just like here, if we roll out our own optional, composing code requires converting from one type to another which may be necessary for an end user if they want to wrap into a guava optional.

n3nash replied:
I like this idea +1

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieDataFile.java:
```diff
@@ -18,53 +18,53 @@
+  private transient FileStatus fileStatus;
+  private final String fileName;
+  private final String fullPathStr;
+  private final long len;
```
s/len/fileLen -> just like you have below in a different class

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieLogFile.java:
```diff
@@ -38,49 +39,53 @@
   public HoodieLogFile(Path logPath) {
     this.path = logPath;
-    this.fileStatus = Optional.empty();
+    this.fileName = logPath == null ? null : logPath.getName();
```
Can we make these optional instead of null checks everywhere ?

n3nash replied:
okay

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieLogFile.java:
```diff
@@ -104,7 +117,7 @@ public HoodieLogFile rollOver(FileSystem fs) throws IOException {
+    }
   }
 
   @Override
```
Let java re-create the hashcode and equals based on all the new members in the object ?

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieLogFile.java:
```diff
@@ -116,16 +129,19 @@ public boolean equals(Object o) {
   @Override
   public String toString() {
-    return "HoodieLogFile {" + path + '}';
+    return "HoodieLogFile{"
```
Same here, don't we want to include fileName or is part of pathStr ?

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/view/LatestFileSliceOnlyFSViewImpl.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * A file-system view implementataion containing only the latest file-slice for each file-group
+ */
+public class LatestFileSliceOnlyFSViewImpl extends HoodieTableFileSystemView implements
```
renaming : HoodieTableFileSystemViewWithLatestFileSlice ?

n3nash replied:
ok!

n3nash commented:
General comment : The HoodieTable needs to hold this cache on the driver. There are 2 repercussions of this : 

1. The hoodieTable object will be much larger not only in the driver but also in the executor (since it needs to be there to avoid listStatus). This means it will eat up part of the JVM memory.

2. The hoodieTable object must be created (and cached) on the driver and sent over to all executors. There is an added network overhead. 

@bvaradar I will share a document with you which shows numbers for each of these. I think it may be worth while to make this whole caching configurable.

@vinothchandar WDYT ?

bvaradar commented:
@vinothchandar : Yet to address your general comment. Will take a look and get back

bvaradar commented:
@vinothchandar @n3nash : Updated commit message with information on #listStatus savings.

Also added a config property to control caching. Wanted to discuss how to measure #listStatus call savings when we try with komondor with caching turned on/off 

Also created a new Option class and replaced guava Optional with that class.

With this, all other comments should have been addressed too.  Please review when you get a chance.

n3nash commented:
@bvaradar The code looks good to me. I just have one concern, caching the listStatus in hoodieTable makes the size of the hoodieTable large. This means more memory use and also when driver needs to send this cache to the executor, there is a minor network overhead along with each executor holding information about all latestFileSlices increase memory overhead in each executor too. I have privately sent you a google doc with all this analysis. I was just wondering, should be keep this cache compressed per partition (or even file id) ? This way when an executor needs the cache, it only decompresses the fileSlice information it needs and the memory overhead is drastically reduced. This will also help in network overhead. The driver memory may remain the same since it needs all FileSLices anyways. There is obviously an additional CPU overhead of compress/decompress. WDYT ?

For eg. 

Cache { 
partition 1 -> compressed_file_slice_info_1
...
...

}

cache.get(partition) -> only decompresses compressed_file_slice_info_1

Just a thought..

bvaradar commented:
Closing this in favor of https://github.com/apache/incubator-hudi/pull/600
